types/valuerenderer/valuerenderer.go

```diff
@@ -0,0 +1,210 @@
+package valuerenderer
+
+import (
+	"errors"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+
+	"golang.org/x/text/language"
+	"golang.org/x/text/message"
+)
+
+// ValueRenderer defines an interface to produce formated output for Int,Dec,Coin types as well as parse a string to Coin or Uint.
+type ValueRenderer interface {
+	Format(interface{}) (string, error)
+	Parse(string) (interface{}, error)
+}
+
+// DefaultValueRenderer defines a struct that implements ValueRenderer interface
+type DefaultValueRenderer struct {
+	// denomToMetadataMap represents a hashmap to lookup banktypes.Metadata by coin denom
+	denomToMetadataMap map[string]banktypes.Metadata
+}
+
+var _ ValueRenderer = &DefaultValueRenderer{}
+
+// TODO consider to move into valuerenderer_test.go
+// NewDefaultValueRenderer  initiates denomToMetadataMap field and returns DefaultValueRenderer struct
+func NewDefaultValueRenderer() DefaultValueRenderer {
+	return DefaultValueRenderer{denomToMetadataMap: make(map[string]banktypes.Metadata)}
+}
+
+// TODO decide what is the most effiecnt way to map key to metadata
+// 1. current implementation O(m*d)
+// 2. use metadata.Symbol is a key ,but this is an optional field
+// 3. use metadata.Base or ,metadata.Display as a key in this case the helper function
+// SetDenomToMetadataMap populates a hashmap that maps coin denomination to a metadata.
+func (dvr DefaultValueRenderer) SetDenomToMetadataMap(metadatas []banktypes.Metadata) error {
+	// TODO should I validate denom?
+	if metadatas == nil {
+		return errors.New("empty metadatas")
+	}
+
+	// O(m*d), wheren m is number of metadatas and d is a number of denomUnits in each metadata
```

Comment:
Do you think this is a good approach to populate `dvr.denomToMetadataMap`? I have outlined some thoughts above L39-L41